virtualbox/vbox-clean-snapshots.py

```diff
@@ -3,49 +3,77 @@
 import sys
 import argparse
 import textwrap
-import virtualbox
-from virtualbox.library import MachineState
-
-
-TO_DELETE = []
-
-
-def get_snapshots_to_delete(snapshot, protected_snapshots):
-    for child in snapshot.children:
-        get_snapshots_to_delete(child, protected_snapshots)
-    snapshot_name = snapshot.name.lower()
-    for protected_str in protected_snapshots:
-        if protected_str.lower() in snapshot_name:
-            return
-    TO_DELETE.append((snapshot.name, snapshot.id_p))
-
+import subprocess
+import re
+from vboxcommon import *
+
+def get_snapshot_children(vm_name, root_snapshot_name, protected_snapshots):
+    """Gets the children of a snapshot using 'VBoxManage showvminfo'.
+    
+    Args:
+      vm_name: The name of the VM.
+      snapshot_name: The name of the snapshot.
+
+    Returns:
+      A list of snapshot names that are children of the given snapshot.
+    """
+    try:
+        vminfo = run_vboxmanage(["showvminfo", vm_name, "--machinereadable"])
+        # Find all snapshot names
+        snapshot_regex = rf'(SnapshotName(?:-\d+)*)=\"(.*?)\"'
+        snapshots = re.findall(snapshot_regex, vminfo, flags=re.M)
+
+        children = []
+
+        # find the root SnapshotName by matching the name
+        root_snapshotid = None
+        for snapshotid, snapshot_name in snapshots:
+            if snapshot_name.lower() == root_snapshot_name.lower() and (not any(p.lower() in snapshot_name.lower() for p in protected_snapshots)):
+                root_snapshotid = snapshotid
+
+        if not root_snapshotid:
+            print("Failed to find root snapshot")
+            raise Exception(f"Failed to find root snapshot {snapshot_name}")
+ 
+        # children of that snapshot share the same prefix id
+        dependant_child = False
+        for snapshotid, snapshot_name in snapshots:
+            if snapshotid.startswith(root_snapshotid):
+                if not any(p.lower() in snapshot_name.lower() for p in protected_snapshots):
+                    children.append((snapshotid, snapshot_name))
+                else:
+                    dependant_child = True
+
+        # remove the root snapshot if any children are protected OR it's the current snapshot
+        if dependant_child:
+            print("Root snapshot cannot be deleted as a child snapshot is protected")
+            children = [snapshot for snapshot in children if snapshot[0] != root_snapshotid]
```

Comment:
i *think* i can follow this logic, and it seems reasonable, but without some example output, i really have any idea.

furthermore, i wouldn't be comfortable changing this logic without any tests. how would i know if i broke anything (especially without vboxmanage installed)?

i'd recommend making this parsing region its own function (input: str, output: list), and then adding some test cases with known data. the test cases will serve both as documentation (since readers can see some examples of real-world data) and to prevent regressions.